[ANDROID] Improve native exception handling to prevent fatal crashes

Crashlytics has been reporting numerous fatal crashes caused by this plugin on Android, for example:

    java.lang.IllegalStateException: 
        at com.smartmobilesoftware.util.IabHelper.flagStartAsync (IabHelper.java:871)
        at com.smartmobilesoftware.util.IabHelper.queryInventoryAsync (IabHelper.java:686)
        at com.smartmobilesoftware.util.IabHelper.queryInventoryAsync (IabHelper.java:714)
        at com.smartmobilesoftware.inappbilling.InAppBillingPlugin$1$1.run (InAppBillingPlugin.java:202)
        at android.os.Handler.handleCallback (Handler.java:873)
        at android.os.Handler.dispatchMessage (Handler.java:99)
        at android.os.Looper.loop (Looper.java:214)
        at android.app.ActivityThread.main (ActivityThread.java:6981)
        at java.lang.reflect.Method.invoke (Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:493)
    at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1445)

<br/>

    java.lang.IllegalStateException: 
        at com.smartmobilesoftware.util.IabHelper.checkSetupDone (IabHelper.java:834)
        at com.smartmobilesoftware.util.IabHelper.queryInventoryAsync (IabHelper.java:685)
        at com.smartmobilesoftware.util.IabHelper.queryInventoryAsync (IabHelper.java:714)
        at com.smartmobilesoftware.inappbilling.InAppBillingPlugin$1$1.run (InAppBillingPlugin.java:202)
        at android.os.Handler.handleCallback (Handler.java:873)
        at android.os.Handler.dispatchMessage (Handler.java:99)
        at android.os.Looper.loop (Looper.java:214)
        at android.app.ActivityThread.main (ActivityThread.java:6981)
        at java.lang.reflect.Method.invoke (Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1445)

<br/>

    java.lang.RuntimeException: 
        at android.app.ActivityThread.performResumeActivity (ActivityThread.java:3788)
        at android.app.ActivityThread.handleResumeActivity (ActivityThread.java:3828)
        at android.app.ActivityThread.handleLaunchActivity (ActivityThread.java:3036)
        at android.app.ActivityThread.-wrap11 (Unknown Source)
        at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1696)
        at android.os.Handler.dispatchMessage (Handler.java:105)
        at android.os.Looper.loop (Looper.java:164)
        at android.app.ActivityThread.main (ActivityThread.java:6938)
        at java.lang.reflect.Method.invoke (Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run (Zygote.java:327)
        at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1374)
        Caused by: java.lang.NullPointerException: 
        at com.smartmobilesoftware.inappbilling.InAppBillingPlugin.onActivityResult (InAppBillingPlugin.java:462)
        at org.apache.cordova.CordovaInterfaceImpl.onActivityResult (CordovaInterfaceImpl.java:152)
        at org.apache.cordova.CordovaActivity.onActivityResult (CordovaActivity.java:358)
        at android.app.Activity.dispatchActivityResult (Activity.java:7539)
        at android.app.ActivityThread.deliverResults (ActivityThread.java:4485)

We should add try/catch handlers in all the native code entry points of the plugin to ensure exceptions are handled gracefully (e.g. reported via error callback or logged to logcat) instead of causing fatal exceptions leading to app crashes.
I'll try to submit a PR to implement this soon.

### dev system info
```
Windows 10 v1809
cordova@8.0.0 / cordova-android@6.4.0
Plugin Version: 8.1.0
```

### example affected devices (from Crashlytics reports)
HWCLT, Android 9
dreamlte, Android 9
star2lte, Android 9
dreamlte, Android 8.0
UMI_MAX, Android 7.0

### Expected behavior
Native exceptions raised within the Android plugin code are handled gracefully and do not lead to fatal exceptions causing app crash.

### Observed behavior
Native exceptions raised within the Android plugin code are not handled gracefully and lead to fatal exceptions causing app crash.

### Steps to reproduce
Cannot give exact steps to reproduce as the crashes we reported via Crashyltics from production apps on user devices.

--- a/src/android/com/smartmobilesoftware/inappbilling/InAppBillingPlugin.java
+++ b/src/android/com/smartmobilesoftware/inappbilling/InAppBillingPlugin.java
@@ -183,31 +183,36 @@ public class InAppBillingPlugin extends CordovaPlugin {
 
                 cordova.getActivity().runOnUiThread(new Runnable() {
                         public void run() {
-
-                            if (!result.isSuccess()) {
-                                // Oh no, there was a problem.
-                                callbackContext.error(IabHelper.ERR_SETUP + "|Problem setting up in-app billing: " + result);
-                                return;
-                            }
-
-                            // Have we been disposed of in the meantime? If so, quit.
-                            if (mHelper == null) {
-                                callbackContext.error(IabHelper.ERR_SETUP + "|The billing helper has been disposed");
-                                return;
-                            }
-
-                            // Hooray, IAB is fully set up. Now, let's get an inventory of stuff we own.
-                            if (skus.size() <= 0) {
-                                Log.d(TAG, "Setup successful. Querying inventory.");
-                                mHelper.queryInventoryAsync(mGotInventoryListener);
-                            }
-                            else {
-                                Log.d(TAG, "Setup successful. Querying inventory w/ SKUs.");
-                                try {
-                                    mHelper.queryInventoryAsync(true, skus, mGotInventoryListener);
-                                } catch (IllegalStateException ex) {
-                                    Log.d("Catch IllegalStateException", ex.getMessage());
+                            try {
+                                if (!result.isSuccess()) {
+                                    // Oh no, there was a problem.
+                                    callbackContext.error(IabHelper.ERR_SETUP + "|Problem setting up in-app billing: " + result);
+                                    return;
                                 }
+
+                                // Have we been disposed of in the meantime? If so, quit.
+                                if (mHelper == null) {
+                                    callbackContext.error(IabHelper.ERR_SETUP + "|The billing helper has been disposed");
+                                    return;
+                                }
+
+                                // Hooray, IAB is fully set up. Now, let's get an inventory of stuff we own.
+                                if (skus.size() <= 0) {
+                                    Log.d(TAG, "Setup successful. Querying inventory.");
+                                    mHelper.queryInventoryAsync(mGotInventoryListener);
+                                }
+                                else {
+                                    Log.d(TAG, "Setup successful. Querying inventory w/ SKUs.");
+                                    try {
+                                        mHelper.queryInventoryAsync(true, skus, mGotInventoryListener);
+                                    } catch (IllegalStateException ex) {
+                                        Log.e(TAG, "Catch IllegalStateException: " + ex.getMessage());
+                                        callbackContext.error(IabHelper.ERR_SETUP + "|" + ex.getMessage());
+                                    }
+                                }
+                            } catch (Exception e) {
+                                Log.e(TAG, e.getMessage());
+                                callbackContext.error(IabHelper.ERR_SETUP + "|" + e.getMessage());
                             }
                         }
                 });
@@ -338,36 +343,45 @@ public class InAppBillingPlugin extends CordovaPlugin {
 	// Listener that's called when we finish querying the items and subscriptions we own
     IabHelper.QueryInventoryFinishedListener mGotInventoryListener = new IabHelper.QueryInventoryFinishedListener() {
         public void onQueryInventoryFinished(IabResult result, Inventory inventory) {
-        	Log.d(TAG, "Inside mGotInventoryListener");
-        	if (hasErrorsAndUpdateInventory(result, inventory)) return;
+            try{
+                Log.d(TAG, "Inside mGotInventoryListener");
+                if (hasErrorsAndUpdateInventory(result, inventory)) return;
 
-            Log.d(TAG, "Query inventory was successful.");
-            callbackContext.success();
-
+                Log.d(TAG, "Query inventory was successful.");
+                callbackContext.success();
+            } catch (Exception e) {
+                Log.e(TAG, e.getMessage());
+                callbackContext.error(IabHelper.ERR_LOAD + "|" + e.getMessage());
+            }
         }
     };
     // Listener that's called when we finish querying the details
     IabHelper.QueryInventoryFinishedListener mGotDetailsListener = new IabHelper.QueryInventoryFinishedListener() {
         public void onQueryInventoryFinished(IabResult result, Inventory inventory) {
-            Log.d(TAG, "Inside mGotDetailsListener");
-            if (hasErrorsAndUpdateInventory(result, inventory)) return;
+            try{
+                Log.d(TAG, "Inside mGotDetailsListener");
+                if (hasErrorsAndUpdateInventory(result, inventory)) return;
 
-            Log.d(TAG, "Query details was successful.");
+                Log.d(TAG, "Query details was successful.");
 
-            List<SkuDetails>skuList = inventory.getAllProducts();
+                List<SkuDetails>skuList = inventory.getAllProducts();
 
-            // Convert the java list to json
-            JSONArray jsonSkuList = new JSONArray();
-            try {
-                for (SkuDetails sku : skuList) {
-                    Log.d(TAG, "SKUDetails: Title: "+sku.getTitle());
-                    jsonSkuList.put(sku.toJson());
+                // Convert the java list to json
+                JSONArray jsonSkuList = new JSONArray();
+                try {
+                    for (SkuDetails sku : skuList) {
+                        Log.d(TAG, "SKUDetails: Title: "+sku.getTitle());
+                        jsonSkuList.put(sku.toJson());
+                    }
+                } catch (JSONException e) {
+                    callbackContext.error(IabHelper.ERR_LOAD + "|" + e.getMessage());
+                    return;
                 }
-            } catch (JSONException e) {
+                callbackContext.success(jsonSkuList);
+            } catch (Exception e) {
+                Log.e(TAG, e.getMessage());
                 callbackContext.error(IabHelper.ERR_LOAD + "|" + e.getMessage());
-		return;
             }
-            callbackContext.success(jsonSkuList);
         }
     };
 
@@ -393,36 +407,41 @@ public class InAppBillingPlugin extends CordovaPlugin {
     // Callback for when a purchase is finished
     IabHelper.OnIabPurchaseFinishedListener mPurchaseFinishedListener = new IabHelper.OnIabPurchaseFinishedListener() {
         public void onIabPurchaseFinished(IabResult result, Purchase purchase) {
-            Log.d(TAG, "Purchase finished: " + result + ", purchase: " + purchase);
+            try{
+                Log.d(TAG, "Purchase finished: " + result + ", purchase: " + purchase);
 
-            // Have we been disposed of in the meantime? If so, quit.
-            if (mHelper == null) {
-            	callbackContext.error(IabHelper.ERR_PURCHASE + "|The billing helper has been disposed");
-            }
+                // Have we been disposed of in the meantime? If so, quit.
+                if (mHelper == null) {
+                    callbackContext.error(IabHelper.ERR_PURCHASE + "|The billing helper has been disposed");
+                }
 
-            if (result.isFailure()) {
-            	callbackContext.error(result.getResponse() + "|Error purchasing: " + result);
-                return;
-            }
+                if (result.isFailure()) {
+                    callbackContext.error(result.getResponse() + "|Error purchasing: " + result);
+                    return;
+                }
 
-            if (!verifyDeveloperPayload(purchase)) {
-            	callbackContext.error(IabHelper.ERR_PURCHASE + "|Error purchasing. Authenticity verification failed.");
-                return;
-            }
+                if (!verifyDeveloperPayload(purchase)) {
+                    callbackContext.error(IabHelper.ERR_PURCHASE + "|Error purchasing. Authenticity verification failed.");
+                    return;
+                }
 
-            Log.d(TAG, "Purchase successful.");
+                Log.d(TAG, "Purchase successful.");
 
-            // add the purchase to the inventory
-            myInventory.addPurchase(purchase);
+                // add the purchase to the inventory
+                myInventory.addPurchase(purchase);
 
-            // append the purchase signature & receipt to the json
-            try {
-                JSONObject purchaseJsonObject = new JSONObject(purchase.getOriginalJson());
-                purchaseJsonObject.put("signature", purchase.getSignature());
-                purchaseJsonObject.put("receipt", purchase.getOriginalJson().toString());
-                callbackContext.success(purchaseJsonObject);
-            } catch (JSONException e) {
-                callbackContext.error(IabHelper.ERR_PURCHASE + "|Could not create JSON object from purchase object");
+                // append the purchase signature & receipt to the json
+                try {
+                    JSONObject purchaseJsonObject = new JSONObject(purchase.getOriginalJson());
+                    purchaseJsonObject.put("signature", purchase.getSignature());
+                    purchaseJsonObject.put("receipt", purchase.getOriginalJson().toString());
+                    callbackContext.success(purchaseJsonObject);
+                } catch (JSONException e) {
+                    callbackContext.error(IabHelper.ERR_PURCHASE + "|Could not create JSON object from purchase object");
+                }
+            } catch (Exception e) {
+                Log.e(TAG, e.getMessage());
+                callbackContext.error(IabHelper.ERR_PURCHASE + "|" + e.getMessage());
             }
 
         }
@@ -431,42 +450,51 @@ public class InAppBillingPlugin extends CordovaPlugin {
     // Called when consumption is complete
     IabHelper.OnConsumeFinishedListener mConsumeFinishedListener = new IabHelper.OnConsumeFinishedListener() {
         public void onConsumeFinished(Purchase purchase, IabResult result) {
-            Log.d(TAG, "Consumption finished. Purchase: " + purchase + ", result: " + result);
+            try{
+                Log.d(TAG, "Consumption finished. Purchase: " + purchase + ", result: " + result);
 
-            // We know this is the "gas" sku because it's the only one we consume,
-            // so we don't check which sku was consumed. If you have more than one
-            // sku, you probably should check...
-            if (result.isSuccess()) {
-                // successfully consumed, so we apply the effects of the item in our
-                // game world's logic
+                // We know this is the "gas" sku because it's the only one we consume,
+                // so we don't check which sku was consumed. If you have more than one
+                // sku, you probably should check...
+                if (result.isSuccess()) {
+                    // successfully consumed, so we apply the effects of the item in our
+                    // game world's logic
 
-                // remove the item from the inventory
-            	myInventory.erasePurchase(purchase.getSku());
-                Log.d(TAG, "Consumption successful. .");
+                    // remove the item from the inventory
+                    myInventory.erasePurchase(purchase.getSku());
+                    Log.d(TAG, "Consumption successful. .");
 
-                callbackContext.success(purchase.getOriginalJson());
+                    callbackContext.success(purchase.getOriginalJson());
 
+                }
+                else {
+                    callbackContext.error(result.getResponse() + "|Error while consuming: " + result);
+                }
+            } catch (Exception e) {
+                Log.e(TAG, e.getMessage());
+                callbackContext.error(IabHelper.ERR_UNKNOWN + "|" + e.getMessage());
             }
-            else {
-                callbackContext.error(result.getResponse() + "|Error while consuming: " + result);
-            }
-
         }
     };
 
     @Override
 	public void onActivityResult(int requestCode, int resultCode, Intent data) {
-        Log.d(TAG, "onActivityResult(" + requestCode + "," + resultCode + "," + data);
+        try{
+            Log.d(TAG, "onActivityResult(" + requestCode + "," + resultCode + "," + data);
 
-        // Pass on the activity result to the helper for handling
-        if (!mHelper.handleActivityResult(requestCode, resultCode, data)) {
-            // not handled, so handle it ourselves (here's where you'd
-            // perform any handling of activity results not related to in-app
-            // billing...
-            super.onActivityResult(requestCode, resultCode, data);
-        }
-        else {
-            Log.d(TAG, "onActivityResult handled by IABUtil.");
+            // Pass on the activity result to the helper for handling
+            if (!mHelper.handleActivityResult(requestCode, resultCode, data)) {
+                // not handled, so handle it ourselves (here's where you'd
+                // perform any handling of activity results not related to in-app
+                // billing...
+                super.onActivityResult(requestCode, resultCode, data);
+            }
+            else {
+                Log.d(TAG, "onActivityResult handled by IABUtil.");
+            }
+        } catch (Exception e) {
+            Log.e(TAG, e.getMessage());
+            callbackContext.error(IabHelper.ERR_UNKNOWN + "|" + e.getMessage());
         }
     }
 

--- a/src/android/com/smartmobilesoftware/util/IabHelper.java
+++ b/src/android/com/smartmobilesoftware/util/IabHelper.java
@@ -509,83 +509,87 @@ public class IabHelper {
         IabResult result;
         if (requestCode != mRequestCode) return false;
 
-        checkNotDisposed();
-        checkSetupDone("handleActivityResult");
+        try{
+            checkNotDisposed();
+            checkSetupDone("handleActivityResult");
 
-        // end of async purchase operation that started on launchPurchaseFlow
-        flagEndAsync();
+            // end of async purchase operation that started on launchPurchaseFlow
+            flagEndAsync();
 
-        if (data == null) {
-            logError("Null data in IAB activity result.");
-            result = new IabResult(ERR_BAD_RESPONSE, "Null data in IAB result");
-            if (mPurchaseListener != null) mPurchaseListener.onIabPurchaseFinished(result, null);
-            return true;
-        }
-
-        int responseCode = getResponseCodeFromIntent(data);
-        String purchaseData = data.getStringExtra(RESPONSE_INAPP_PURCHASE_DATA);
-        String dataSignature = data.getStringExtra(RESPONSE_INAPP_SIGNATURE);
-
-        if (resultCode == Activity.RESULT_OK && responseCode == BILLING_RESPONSE_RESULT_OK) {
-            logDebug("Successful resultcode from purchase activity.");
-            logDebug(" - Billing key: " + mSignatureBase64);
-            logDebug(" - Purchase data: " + purchaseData);
-            logDebug(" - Data signature: " + dataSignature);
-            logDebug(" - Extras: " + data.getExtras());
-            logDebug(" - Expected item type: " + mPurchasingItemType);
-
-            if (purchaseData == null || dataSignature == null) {
-                logError("BUG: either purchaseData or dataSignature is null.");
-                logDebug("Extras: " + data.getExtras().toString());
-                result = new IabResult(ERR_UNKNOWN, "IAB returned null purchaseData or dataSignature");
+            if (data == null) {
+                logError("Null data in IAB activity result.");
+                result = new IabResult(ERR_BAD_RESPONSE, "Null data in IAB result");
                 if (mPurchaseListener != null) mPurchaseListener.onIabPurchaseFinished(result, null);
                 return true;
             }
 
-            Purchase purchase = null;
-            try {
-                purchase = new Purchase(mPurchasingItemType, purchaseData, dataSignature);
-                String sku = purchase.getSku();
+            int responseCode = getResponseCodeFromIntent(data);
+            String purchaseData = data.getStringExtra(RESPONSE_INAPP_PURCHASE_DATA);
+            String dataSignature = data.getStringExtra(RESPONSE_INAPP_SIGNATURE);
 
-                // Verify signature
-                if (!Security.verifyPurchase(mSignatureBase64, purchaseData, dataSignature)) {
-                    logError("Purchase signature verification FAILED for sku " + sku);
-                    result = new IabResult(ERR_VERIFICATION_FAILED, "Signature verification failed for sku " + sku);
-                    if (mPurchaseListener != null) mPurchaseListener.onIabPurchaseFinished(result, purchase);
+            if (resultCode == Activity.RESULT_OK && responseCode == BILLING_RESPONSE_RESULT_OK) {
+                logDebug("Successful resultcode from purchase activity.");
+                logDebug(" - Billing key: " + mSignatureBase64);
+                logDebug(" - Purchase data: " + purchaseData);
+                logDebug(" - Data signature: " + dataSignature);
+                logDebug(" - Extras: " + data.getExtras());
+                logDebug(" - Expected item type: " + mPurchasingItemType);
+
+                if (purchaseData == null || dataSignature == null) {
+                    logError("BUG: either purchaseData or dataSignature is null.");
+                    logDebug("Extras: " + data.getExtras().toString());
+                    result = new IabResult(ERR_UNKNOWN, "IAB returned null purchaseData or dataSignature");
+                    if (mPurchaseListener != null) mPurchaseListener.onIabPurchaseFinished(result, null);
                     return true;
                 }
-                logDebug("Purchase signature successfully verified.");
-            }
-            catch (JSONException e) {
-                logError("Failed to parse purchase data.");
-                e.printStackTrace();
-                result = new IabResult(ERR_BAD_RESPONSE, "Failed to parse purchase data.");
-                if (mPurchaseListener != null) mPurchaseListener.onIabPurchaseFinished(result, null);
-                return true;
-            }
 
-            if (mPurchaseListener != null) {
-                mPurchaseListener.onIabPurchaseFinished(new IabResult(BILLING_RESPONSE_RESULT_OK, "Success"), purchase);
+                Purchase purchase = null;
+                try {
+                    purchase = new Purchase(mPurchasingItemType, purchaseData, dataSignature);
+                    String sku = purchase.getSku();
+
+                    // Verify signature
+                    if (!Security.verifyPurchase(mSignatureBase64, purchaseData, dataSignature)) {
+                        logError("Purchase signature verification FAILED for sku " + sku);
+                        result = new IabResult(ERR_VERIFICATION_FAILED, "Signature verification failed for sku " + sku);
+                        if (mPurchaseListener != null) mPurchaseListener.onIabPurchaseFinished(result, purchase);
+                        return true;
+                    }
+                    logDebug("Purchase signature successfully verified.");
+                }
+                catch (JSONException e) {
+                    logError("Failed to parse purchase data.");
+                    e.printStackTrace();
+                    result = new IabResult(ERR_BAD_RESPONSE, "Failed to parse purchase data.");
+                    if (mPurchaseListener != null) mPurchaseListener.onIabPurchaseFinished(result, null);
+                    return true;
+                }
+
+                if (mPurchaseListener != null) {
+                    mPurchaseListener.onIabPurchaseFinished(new IabResult(BILLING_RESPONSE_RESULT_OK, "Success"), purchase);
+                }
             }
-        }
-        else if (resultCode == Activity.RESULT_OK) {
-            // result code was OK, but in-app billing response was not OK.
-            logDebug("Result code was OK but in-app billing response was not OK: " + getResponseDesc(responseCode));
-            if (mPurchaseListener != null) {
-                result = new IabResult(responseCode, "Problem purchashing item.");
-                mPurchaseListener.onIabPurchaseFinished(result, null);
+            else if (resultCode == Activity.RESULT_OK) {
+                // result code was OK, but in-app billing response was not OK.
+                logDebug("Result code was OK but in-app billing response was not OK: " + getResponseDesc(responseCode));
+                if (mPurchaseListener != null) {
+                    result = new IabResult(responseCode, "Problem purchashing item.");
+                    mPurchaseListener.onIabPurchaseFinished(result, null);
+                }
             }
-        }
-        else if (resultCode == Activity.RESULT_CANCELED) {
-            logDebug("Purchase canceled - Response: " + getResponseDesc(responseCode));
-            result = new IabResult(ERR_CANCELLED, "User canceled.");
-            if (mPurchaseListener != null) mPurchaseListener.onIabPurchaseFinished(result, null);
-        }
-        else {
-            logError("Purchase failed. Result code: " + Integer.toString(resultCode)
-                    + ". Response: " + getResponseDesc(responseCode));
-            result = new IabResult(ERR_UNKNOWN, "Unknown purchase response.");
-            if (mPurchaseListener != null) mPurchaseListener.onIabPurchaseFinished(result, null);
+            else if (resultCode == Activity.RESULT_CANCELED) {
+                logDebug("Purchase canceled - Response: " + getResponseDesc(responseCode));
+                result = new IabResult(ERR_CANCELLED, "User canceled.");
+                if (mPurchaseListener != null) mPurchaseListener.onIabPurchaseFinished(result, null);
+            }
+            else {
+                logError("Purchase failed. Result code: " + Integer.toString(resultCode)
+                        + ". Response: " + getResponseDesc(responseCode));
+                result = new IabResult(ERR_UNKNOWN, "Unknown purchase response.");
+                if (mPurchaseListener != null) mPurchaseListener.onIabPurchaseFinished(result, null);
+            }
+        } catch (Exception e) {
+            Log.e("IabHelper", e.getMessage());
         }
         return true;
     }
@@ -702,7 +706,11 @@ public class IabHelper {
                 if (!mDisposed && listener != null) {
                     handler.post(new Runnable() {
                         public void run() {
-                            listener.onQueryInventoryFinished(result_f, inv_f);
+                            try{
+                                listener.onQueryInventoryFinished(result_f, inv_f);
+                            } catch (Exception e) {
+                                Log.e("IabHelper", e.getMessage());
+                            }
                         }
                     });
                 }
@@ -970,7 +978,7 @@ public class IabHelper {
           return ERR_UNKNOWN;
         }
 
-        // Split the SKUs into slices of maximum 20 entries before querying them to prevent 
+        // Split the SKUs into slices of maximum 20 entries before querying them to prevent
         // "Input Error: skusBundle array associated with key ITEM_ID_LIST cannot contain more than 20 items."
         while (skuList.size() > 0) {
             ArrayList<String> skuSubList = new ArrayList<String>(
@@ -1028,31 +1036,35 @@ public class IabHelper {
         flagStartAsync("consume");
         (new Thread(new Runnable() {
             public void run() {
-                final List<IabResult> results = new ArrayList<IabResult>();
-                for (Purchase purchase : purchases) {
-                    try {
-                        consume(purchase);
-                        results.add(new IabResult(BILLING_RESPONSE_RESULT_OK, "Successful consume of sku " + purchase.getSku()));
+                try{
+                    final List<IabResult> results = new ArrayList<IabResult>();
+                    for (Purchase purchase : purchases) {
+                        try {
+                            consume(purchase);
+                            results.add(new IabResult(BILLING_RESPONSE_RESULT_OK, "Successful consume of sku " + purchase.getSku()));
+                        }
+                        catch (IabException ex) {
+                            results.add(ex.getResult());
+                        }
                     }
-                    catch (IabException ex) {
-                        results.add(ex.getResult());
-                    }
-                }
 
-                flagEndAsync();
-                if (!mDisposed && singleListener != null) {
-                    handler.post(new Runnable() {
-                        public void run() {
-                            singleListener.onConsumeFinished(purchases.get(0), results.get(0));
-                        }
-                    });
-                }
-                if (!mDisposed && multiListener != null) {
-                    handler.post(new Runnable() {
-                        public void run() {
-                            multiListener.onConsumeMultiFinished(purchases, results);
-                        }
-                    });
+                    flagEndAsync();
+                    if (!mDisposed && singleListener != null) {
+                        handler.post(new Runnable() {
+                            public void run() {
+                                singleListener.onConsumeFinished(purchases.get(0), results.get(0));
+                            }
+                        });
+                    }
+                    if (!mDisposed && multiListener != null) {
+                        handler.post(new Runnable() {
+                            public void run() {
+                                multiListener.onConsumeMultiFinished(purchases, results);
+                            }
+                        });
+                    }
+                } catch (Exception e) {
+                    Log.e("IabHelper", e.getMessage());
                 }
             }
         })).start();

